System/Directory/Template.hs

```diff
@@ -0,0 +1,1561 @@
+
```

Comment:
```hs
class Path p where
  toAbstractPath :: p -> AbstractFilePath
  fromAbstractPath :: AbstractFilePath -> p
```

this btw won't work well, because the conversion from `String` to `AbstractFilePath` is not total. String can represent invalid/incomplete surrogate pairs.

E.g. try

```
> GHC.withCString utf8 "\55996" BS.packCString
*** Exception: recoverEncode: invalid argument (invalid character)
```

Reply:
So I added some functions that should make your approach possible:

* https://hackage.haskell.org/package/filepath-1.4.99.0/candidate/docs/System-OsPath-Windows.html#v:fromPlatformStringFS
* https://hackage.haskell.org/package/filepath-1.4.99.0/candidate/docs/System-OsPath-Windows.html#v:toPlatformStringFS

These mimic exactly what base library does on the outer FFI layer:

1. on unix shady PEP 383 roundtripping
2. on windows some private surrogate escaping mechanism

So these functions can be used to convert to and from `WindowsPath` to `FilePath`. They'll fail the same way as the base functions if weird encoding is set. These run in IO, because it requires a call to `getFileSystemEncoding` on unix.